src/main/java/com/github/knaufk/flink/faker/FlinkFakerTableSourceFactory.java

```diff
@@ -69,25 +77,52 @@ public FlinkFakerTableSource createDynamicTableSource(final Context context) {
     TableSchema schema = TableSchemaUtils.getPhysicalSchema(catalogTable.getSchema());
     Float[] fieldNullRates = new Float[schema.getFieldCount()];
     String[] fieldExpressions = new String[schema.getFieldCount()];
+    Integer[] fieldArrayLengths = new Integer[schema.getFieldCount()];
 
     for (int i = 0; i < fieldExpressions.length; i++) {
       String fieldName = schema.getFieldName(i).get();
       DataType dataType = schema.getFieldDataType(i).get();
       validateDataType(fieldName, dataType);
 
-      fieldExpressions[i] = readAndValidateFieldExpression(options, fieldName);
-      ;
+      fieldExpressions[i] = readAndValidateFieldExpression(options, fieldName, dataType);
       fieldNullRates[i] = readAndValidateNullRate(options, fieldName);
-      ;
+      fieldArrayLengths[i] = readAndValidateCollectionLength(options, fieldName, dataType);
     }
     return new FlinkFakerTableSource(
         fieldExpressions,
         fieldNullRates,
+        fieldArrayLengths,
         schema,
         options.get(ROWS_PER_SECOND),
         options.get(NUMBER_OF_ROWS));
   }
 
+  private Integer readAndValidateCollectionLength(
+      Configuration options, String fieldName, DataType dataType) {
+    ConfigOption<Integer> collectionLength =
+        key(FIELDS + "." + fieldName + "." + COLLECTION_LENGTH).intType().defaultValue(1);
+    Integer fieldArrayLength = options.get(collectionLength);
+
+    if (fieldArrayLength != 1
+        && !COLLECTION_ROOT_TYPES.contains(dataType.getLogicalType().getTypeRoot())) {
+      throw new ValidationException(
+          "Collection length may not be set for  "
+              + fieldName
+              + " with type "
+              + dataType.getLogicalType().getTypeRoot().toString());
+    }
+
+    if (fieldArrayLength <= 0
+        && COLLECTION_ROOT_TYPES.contains(dataType.getLogicalType().getTypeRoot())) {
+      throw new ValidationException(
+          "Collection length needs to be in positive. Collection length of "
```

Comment:
Thank you for your useful remarks and comments, I will work on them. I also try to come up with a general solution instead of the string delimiter problem. I check whether the MULTISET can be added to the solution with the same logic as the others too. I plan to submit my new solutions next week.